# Hola Cesar!

Mi nombre es David Bautista, soy code reviewer de Tripleten y hoy tengo el gusto de revisar tu proyecto.

Cuando vea un error laa primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión.

Encontrarás mis comentarios más abajo - por favor, no los muevas, no los modifiques ni los borres.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General
    
Hola Cesar, te felicito por el desarrollo del proyecto, todos los ejercicios los desarrollaste de manera correcta y con mucha calidad. Muy buen trabajo. He dejado comentarios para que los puedas revisar y retroalimentar.   </div>

# Tabla de Contenido
0. Introducción<br />

1. Descarga de datos y preparación para el análisis<br />

2. Estudios de Tablas<br />
 2.1 Tabla "books"<br />
 2.2 Tabla "authors"<br />
 2.3 Tabla "publishers"<br />
 2.4 Tabla "ratings"<br />
 2.5 Tabla "reviews"<br />

3. Ejercicio<br />
 3.1 Encuentra el número de libros publicados después del 1 de enero de 2000<br />
 3.2 Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro<br />
 3.3 Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas<br />
 3.4 Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones<br />
 3.5 Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo estructurando la tabla de contenido.  </div>

# Introducción

El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

Te han dado una base de datos de uno de los servicios que compiten en este mercado. Contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros. Esta información se utilizará para generar una propuesta de valor para un nuevo producto.

Ejercicio: 

- Encuentra el número de libros publicados después del 1 de enero de 2000.
- Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.
- Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).
- Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.
- Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo generando la introdcción del proyecto.  </div>

## Descarga de datos y preparaciónpara el análisis.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo importando las librerías necesarias para el desarrollo del proyecto. </div>

In [2]:
db_config = {'user': 'practicum_student',         
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs',
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,             
             'db': 'data-analyst-final-project-db'}         

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, Cesar, buen trabajo configurando la conexión al servidor SQL.  </div>

## Estudios de Tablas

### Tabla "books"

In [3]:
query = """
SELECT *
FROM books
limit 5
"""

In [4]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Tabla "authors"

In [5]:
query = """
SELECT *
FROM authors
limit 5
"""

In [6]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### Tabla "publishers"

In [7]:
query = """
SELECT *
FROM publishers
limit 5
"""

In [8]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Tabla "ratings"

In [9]:
query = """
SELECT *
FROM ratings
limit 5
"""

In [10]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### Tabla "reviews"

In [11]:
query = """
SELECT *
FROM reviews
limit 5
"""

In [12]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Genial, buen trabajo con el desarrollo de esta pequeña exploración inicial de los datos por medio del uso de sentencias ``SQL``.   </div>

## Ejercicio

### Encuentra el número de libros publicados después del 1 de enero de 2000.

In [13]:
query_ejercicio1 = """
SELECT COUNT(*) as count_books
FROM books
WHERE publication_date > '2000-01-01';  
"""
results = pd.io.sql.read_sql(query_ejercicio1, con = engine)
display(results)

count_books
0          819

- Se encontró un total de 819 libros que fueron publicados con posterioridad al 1 de enero de 2000.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo, identificando la cantidad de libros publicados después del 1 de enero del 2000.

### Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [14]:
query_ejercicio2 = """
SELECT books.book_id, 
    authors.author,
    books.title, 
    COUNT(distinct review_id) as number_users_reviews, 
    ROUND(AVG(ratings.rating),2) as average_rating
FROM books
    left join reviews on books.book_id = reviews.book_id 
    left join ratings on books.book_id = ratings.book_id 
    left join authors on books.author_id = authors.author_id 
group by books.book_id, authors.author_id
order by average_rating desc 
"""
results = pd.io.sql.read_sql(query_ejercicio2, con = engine)
display(results[:5])

book_id                                             author  \
0       86                                    Mercedes Lackey   
1      901  Robert Kirkman/Tony Moore/Charlie Adlard/Cliff...   
2      390                                   William Faulkner   
3      972                                     Jon Kabat-Zinn   
4      136                       John Eldredge/Stasi Eldredge   

                                               title  number_users_reviews  \
0      Arrows of the Queen (Heralds of Valdemar  #1)                     2   
1  The Walking Dead  Book One (The Walking Dead #...                     2   
2                                    Light in August                     2   
3  Wherever You Go  There You Are: Mindfulness Me...                     2   
4  Captivating: Unveiling the Mystery of a Woman'...                     2   

   average_rating  
0             5.0  
1             5.0  
2             5.0  
3             5.0  
4             5.0

- Tras procesar los datos, se elaboró una tabla que resume el conteo de reseñas por cada libro y la calificación promedio obtenida.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Excelente, buen trabajo estructurando la tabla.

###  Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas

In [15]:
query_ejercicio3 = """
select 
    distinct publishers.publisher_id 
    , publishers.publisher
    , COUNT(books.book_id) OVER(PARTITION BY publishers.publisher_id) AS count_books 
from books 
    inner join publishers using(publisher_id)
where books.num_pages > 50
order by count_books desc
limit 1
"""
results = pd.io.sql.read_sql(query_ejercicio3, con = engine)
display(results[:5])

publisher_id      publisher  count_books
0           212  Penguin Books           42

- Al analizar, nos damos cuenta que el sello editorial Penguin Books resultó ser el que mayor cantidad de libros ha publicado.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, ``Penguin Books`` es la editorial que ha publicado el mayor número de libros de mas de 50 paginas.

###  Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [16]:
query_ejercicio4 = """
with books_rating(book_id,title,author,count_marks,avg_rating)
as (
select 
   distinct book_id
   , books.title 
   , authors.author 
   , count(rating_id) over (partition by book_id) as count_marks
   , round(avg(rating) over (partition by book_id),2) as avg_rating
from books 
    left join ratings using(book_id)
    left join authors using(author_id)
)
select 
    book_id
    , title
    , author
    , avg_rating
from books_rating
where 
    (count_marks > 50) and avg_rating = 
       (
        select avg_rating 
        from books_rating
        where count_marks > 50
        order by avg_rating desc
        limit 1
        )
"""
results = pd.io.sql.read_sql(query_ejercicio4, con = engine)
display(results[:5])

book_id                                              title  \
0      302  Harry Potter and the Prisoner of Azkaban (Harr...   

                       author  avg_rating  
0  J.K. Rowling/Mary GrandPré        4.41

- El autor con la puntuación media más alta

In [17]:
query_ejercicio4_1 = """
with books_rating(book_id,author_id, count_marks,avg_rating)
as (
select 
   distinct book_id
   , authors.author_id 
   , count(rating_id) over (partition by book_id) as count_marks
   , round(avg(rating) over (partition by book_id),2) as avg_rating
from books 
   left join ratings using(book_id)
   right join authors using(author_id)
)
select
    authors.author_id
    , authors.author 
    , COUNT(book_id) as count_books
    , AVG(avg_rating) as avg_rat    
from books_rating
inner join authors on books_rating.author_id = authors.author_id 
where count_marks > 50
group by authors.author_id
order by avg_rat desc 
limit 1
"""
results = pd.io.sql.read_sql(query_ejercicio4_1, con = engine)
display(results[:5])

author_id                      author  count_books  avg_rat
0        236  J.K. Rowling/Mary GrandPré            4    4.285

- Se arrojó un resultado doble sobre la autora J.K. Rowling. Por un lado, se identificó que su libro es el mejor calificado. Por otro lado, se determinó que, en general, los libros de J.K. Rowling poseen las calificaciones más altas.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Excelente, buen trabajo identificando a la autora y así mismo con el desarrollo adicional para identificar que por lo general los libros de ``J.K`` poseen calificaciones altas.  </div>

### Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [18]:
query_ejercicio5 = """ 
with users_rated_50_plus (username)
as (
select 
   ratings.username 
from ratings 
group by username
having count(rating_id) > 50
),
users_count_reviews (username, count_reviews)
as (
    SELECT 
        distinct username
        , COUNT(review_id) over (partition by username) as count_reviews
    from users_rated_50_plus
        inner join reviews USING(username)
)
select ROUND(AVG(count_reviews),2) as AVG_REVIEWS 
from users_count_reviews
"""
results = pd.io.sql.read_sql(query_ejercicio5, con = engine)
display(results[:5])

avg_reviews
0        24.33

- Nuestro análisis arrojó un promedio de 24.33 reseñas por usuario.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, muy buen trabajo realizando el cálculo.  </div>